modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Resources.scala

```diff
@@ -0,0 +1,242 @@
+    projectId: String,
+    topic: String,
+    maxConcurrency: Int,
+    metrics: Metrics[F],
+    sinkSettingsBad: SinkSettings.Bad
+  ): Resource[F, Pipe[F, StreamBadRow[F], Unit]] =
+    mkProducer[F, BadRow](
+      projectId,
+      topic,
+      sinkSettingsBad.producerBatchSize,
+      sinkSettingsBad.producerDelayThreshold
+    ).map { p =>
+      _.parEvalMapUnordered(maxConcurrency) { badRow =>
+        p.produce(badRow.row) *> badRow.ack *> metrics.badCount
+      }
+    }
+
+  // Does not ack the event -- it still needs to end up in one of the other targets
+  private def mkTypeSink[F[_]: Concurrent](
+    projectId: String,
+    topic: String,
+    sinkSettingsTypes: SinkSettings.Types
+  ): Resource[F, Pipe[F, Set[ShreddedType], Unit]] =
+    mkProducer[F, Set[ShreddedType]](
+      projectId,
+      topic,
+      sinkSettingsTypes.producerBatchSize,
+      sinkSettingsTypes.producerDelayThreshold
+    ).map { p =>
+      _.parEvalMapUnordered(sinkSettingsTypes.sinkConcurrency) { types =>
+        p.produce(types).void
+      }
+    }
+
+  /** Write good events through a BigQuery sink and pipe observed types through a Pub/Sub sink.
+    * Acks the events.
+    *
+    * The input Stream of events is transformed into a Stream of batches of events (List[StreamLoaderRow]).
+    * Each batch respects the Streaming APIs recommended number of records per request.
+    * Each BQ write request also has a size limit of 10MB. In case a batch's total size exceeds this limit, we use
+    * as many events as possible (while staying below it) in the write request, and re-insert the remaining events
+    * back into the original Stream of events.
+    *
+    * @param bufferOverflowQueue A Queue in which put events that are left over from a batch after taking as many as
+    *                            possible without breaching the streaming inserts request size limit
+    */
+  private def mkGoodSink[F[_]: Concurrent: ContextShift: Timer](
+    blocker: Blocker,
+    good: Output.BigQuery,
+    bigQuery: BigQuery,
+    producer: PubsubProducer[F, FailedInsert],
+    metrics: Metrics[F],
+    typeSink: Pipe[F, Set[ShreddedType], Unit],
+    bufferOverflowQueue: Queue[F, StreamLoaderRow[F]],
+    sinkSettingsGood: SinkSettings.Good,
+    sinkSettingsTypes: SinkSettings.Types
+  ): Pipe[F, StreamLoaderRow[F], Unit] = {
+    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = {
+      _.merge(bufferOverflowQueue.dequeue)
+        .through(batch(sinkSettingsGood.bqWriteRequestThreshold, sinkSettingsGood.bqWriteRequestTimeout))
+        .parEvalMapUnordered(sinkSettingsGood.sinkConcurrency) { slrows =>
+          val (remaining, toInsert) = splitBy(slrows, getSize[F], sinkSettingsGood.bqWriteRequestSizeLimit)
+          remaining.traverse_(bufferOverflowQueue.enqueue1) *>
+            blocker.blockOn(
+              Bigquery.insert(producer, metrics, toInsert.map(_.row))(Bigquery.mkInsert(good, bigQuery)) *> toInsert
```

Comment:
Ideally you would run only the truly blocking call on the blocker, and return to the default execution context as quickly as possible.  I think in this case it needs to look like this:

```
  remaining.traverse_(bufferOverflowQueue.enqueue1) *>
    Bigquery.insert(producer, metrics, toInsert.map(_.row))(blocker.blockOn(Bigquery.mkInsert(good, bigQuery))) *> toInsert
    .traverse_(_.ack)
```

Reply:
Sorry, I'm not sure I understand this suggestion.

`mkInsert` just produces a function `List[LoaderRow] => F[InsertAllResponse]`. It is `BigQuery.insert` that does the blocking work. Also, do we not want to have `ack` on the blocker as well?

Reply:
`mkInsert` is a wrapper around `bigQuery.insertAll(request)` ([proof](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/1.0.0-rc3/modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Bigquery.scala#L58)), which is blocking.  So the `blockOn` only needs to wrap the call to `Bigquery.mkInsert`, nothing else.

Even better, you could consider re-defining `mkInsert` so it takes a `Blocker` as an extra parameter.  I like this because it is most explicit to someone reading the code about which call is blocking.  And it prevents attempted mis-use of `mkInsert` without using any blocker at all (imagine in a future code refactor).

```scala
  def mkInsert[F[_]: Sync: ContextShift](good: Output.BigQuery, bigQuery: BigQuery, blocker: Blocker): List[LoaderRow] => F[InsertAllResponse] = {
    lrs =>
      val request = buildRequest(good.datasetId, good.tableId, lrs)
      blocker.delay(bigQuery.insertAll(request))
```

You don't need a blocker for the `ack`, because the underlying pubsub libraries do the work on a different thread pool already.

Reply:
> So the `blockOn` only needs to wrap the call to `Bigquery.mkInsert`, nothing else.

Just for my understanding, isn't `BigQuery.mkInsert` just a value that is being passed around until we try to do something with the response?

Reply:
Oh sorry you are right, I put the blocker in the wrong place in my first example (original comment).  I completely confused you with that.

But I still stand by my other points:
- Wrap the minimum possible code in a blocker.  Apart from the truly blocking calls, try to run as much as possible on the default thread pool
- PubSub acks do not block our threads, because of how they are implemented in underlying libraries using SettableFutures, which run on a different thread pool ([here if you are really interested](https://github.com/googleapis/java-pubsub/blob/be6a48fc83d15a320bc86a8b64dc55e66ad6ac11/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java#L370-L372))

I think my code example in my second comment is the best place to use the blocker.